vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,391 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// Package proxyapp package implements the experimental plugins support.
+// We promise interface part will not be stable until documented.
+package proxyapp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/rpc"
+	"net/rpc/jsonrpc"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/vm/proxyapp/proxyrpc"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func ctor(params *proxyAppParams, env *vmimpl.Env) (vmimpl.Pool, error) {
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	p := &pool{
+		env:      env,
+		close:    make(chan bool, 1),
+		onClosed: make(chan error, 1),
+	}
+
+	err = p.init(params, subConfig)
+	if err != nil {
+		return nil, fmt.Errorf("can't initialize pool: %w", err)
+	}
+	needTerminate := true
+
+	go func() {
+		for {
+			select {
+			case <-p.close:
+				if needTerminate {
+					p.proxy.terminate()
+					<-p.proxy.onTerminated
+				}
+				p.onClosed <- nil
+				return
+			case <-p.proxy.onTerminated:
+			}
+			time.Sleep(params.InitRetryDelay)
+			needTerminate = nil != p.init(params, subConfig)
+		}
+	}()
+
+	return p, nil
+}
+
+type pool struct {
+	mu       sync.Mutex
+	env      *vmimpl.Env
+	proxy    *ProxyApp
+	count    int
+	close    chan bool
+	onClosed chan error
+}
+
+func (p *pool) init(params *proxyAppParams, cfg *Config) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	var err error
+	p.proxy, err = runProxyApp(params, cfg.Command)
+	if err != nil {
+		return fmt.Errorf("failed to run ProxyApp: %w", err)
+	}
+
+	count, err := p.proxy.CreatePool(string(cfg.ProxyAppConfig), p.env.Debug)
+	if err != nil || count == 0 || (p.count != 0 && p.count != count) {
+		if err == nil {
+			err = fmt.Errorf("wrong pool size %v, prev was %v", count, p.count)
+		}
+
+		p.proxy.terminate()
+		<-p.proxy.onTerminated
+		p.proxy = nil
+
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	if p.count == 0 {
+		p.count = count
+	}
+	return nil
+}
+
+func (p *pool) Count() int {
+	return p.count
+}
+
+func (p *pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	p.mu.Lock()
+	proxy := p.proxy
+	p.mu.Unlock()
+
+	return proxy.CreateInstance(workdir, index)
```

Comment:
proxy can be nil. Add a test that exposes this.

Reply:
Added test.
Additionally moved pool.mu.lock() out of init() func.
Want Create() to be blocked while "time.Sleep(params.InitRetryDelay)".